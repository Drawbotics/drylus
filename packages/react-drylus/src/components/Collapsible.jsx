import React from 'react';
import { css } from 'emotion';
import PropTypes from 'prop-types';
import sv from '@drawbotics/style-vars';

import Label from './Label';
import Icon from './Icon';


const styles = {
  root: css`
  `,
  header: css`
    display: flex;
    align-items: center;
    justify-content: space-between;

    & [data-element="title"] {
      transition: ${sv.defaultTransition};
    }

    &:hover {
      cursor: pointer;

      & [data-element="title"] {
        transform: translateX(5px);
      }
    }
  `,
  icon: css`
    margin-bottom: -2px;
    color: ${sv.colorSecondary};
  `,
  content: css`
    padding-top: ${sv.paddingExtraSmall};
  `,
};


const Collapsible = ({
  title,
  isOpen,
  children,
  onClick,
}) => {
  return (
    <div className={styles.root} onClick={onClick}>
      <div className={styles.header}>
        <div data-element="title">
          <Label ellipsized>{title}</Label>
        </div>
        <div className={styles.icon}>
          <Icon name={isOpen ? 'chevron-up' : 'chevron-down'} />
        </div>
      </div>
      {do {
        if (isOpen) {
          <div className={styles.content}>
            {children}
          </div>
        }
      }}
    </div>
  );
};


Collapsible.propTypes = {
  /** Shown on the left side as a label */
  title: PropTypes.string,

  /** Determines whether the content of the collapsible is visible */
  isOpen: PropTypes.bool,

  /** The togglable content */
  children: PropTypes.node.isRequired,

  /** Triggered when the arrow is clicked */
  onClick: PropTypes.func,
};


export default Collapsible;
