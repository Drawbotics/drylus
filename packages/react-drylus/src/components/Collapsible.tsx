import sv from '@drawbotics/drylus-style-vars';
import { css } from 'emotion';
import get from 'lodash/get';
import React, { ReactNode } from 'react';

import { Shade } from '../enums';
import { ListTile } from '../layout';
import { OnClickCallback, Style } from '../types';
import { Deprecated, checkComponentProps, run } from '../utils';
import { Icon, IconType } from './Icon';
import { Label } from './Label';

const styles = {
  root: css``,
  header: css`
    display: flex;
    align-items: center;

    & [data-element='title'] {
      transition: ${sv.transitionShort};
      flex: 1;

      /* flex item */
      > div > div > div {
        display: flex;
      }
    }

    &:hover {
      cursor: pointer;

      & [data-element='title'] {
        transform: translateX(5px);
      }
    }
  `,
  arrow: css`
    margin-bottom: -2px;
    color: ${sv.colorSecondary};
  `,
  content: css`
    padding-top: ${sv.paddingExtraSmall};
  `,
  trailing: css`
    margin-left: ${sv.paddingExtraSmall};
  `,
};

export interface CollapsibleProps {
  /** Shown on the left side as a label */
  title: React.ReactNode;

  /** Determines whether the content of the collapsible is visible */
  isOpen?: boolean;

  /** The togglable content */
  children: React.ReactNode;

  /** Triggered when the arrow is clicked */
  onClick?: OnClickCallback<HTMLDivElement>;

  /**
   * If given displays an icon to the left of the title
   * @deprecated Use leading instead
   */
  icon?: IconType;

  /** If given, renders in front of the collapsible title. For now limited to Icon */
  leading?: React.ReactElement<typeof Icon> | ReactNode;

  /** If given, renders after the collapsible arrow. Can be anything */
  trailing?: ReactNode;

  /** Used for style overrides */
  style?: Style;
}

export const Collapsible = ({
  title,
  isOpen,
  children,
  onClick,
  style,
  icon,
  leading: _leading,
  trailing,
}: CollapsibleProps) => {
  checkComponentProps({ leading: _leading }, { leading: [Icon] });

  const leading =
    _leading != null ? _leading : icon != null ? <Icon name={icon} shade={Shade.LIGHT} /> : null;
  const isValidLeading = get(leading, 'type') === Icon;

  return (
    <div style={style} className={styles.root}>
      <div className={styles.header}>
        <div data-element="title" onClick={onClick}>
          <ListTile
            title={typeof title === 'string' ? <Label ellipsized>{title}</Label> : title}
            leading={isValidLeading ? leading : null}
          />
        </div>
        <div className={styles.arrow} onClick={onClick}>
          <Icon name={isOpen ? 'chevron-up' : 'chevron-down'} />
        </div>
        {trailing != null ? <div className={styles.trailing}>{trailing}</div> : null}
      </div>
      {run(() => {
        if (isOpen) {
          return <div className={styles.content}>{children}</div>;
        }
      })}
    </div>
  );
};

Collapsible.propTypes = {
  icon: Deprecated,
};
