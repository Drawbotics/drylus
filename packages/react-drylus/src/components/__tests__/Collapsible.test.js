import React from 'react';
import { create } from 'react-test-renderer';

import Collapsible from '../Collapsible';


describe('Collapsible', () => {
  describe('matches snapshot when', () => {
    it('is not open', () => {
      const tree = create(
        <Collapsible title="Title" isOpen={false}>Collapsible content</Collapsible>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('it is open', () => {
      const tree = create(
        <Collapsible title="Title" isOpen={true}>Collapsible content</Collapsible>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('opens when clicked', () => {
    let open = false;

    expect(open).toBeFalsy();

    const component = create(
      <Collapsible
        title="Title"
        isOpen={false}
        onClick={() => open = true}>
        Collapsible content
      </Collapsible>
    );

    component.root.props.onClick();
    
    expect(open).toBeTruthy();
  });
});