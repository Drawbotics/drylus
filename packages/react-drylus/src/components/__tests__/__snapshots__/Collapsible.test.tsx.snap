// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collapsible matches snapshot when is not open 1`] = `
.emotion-4 {
  color: inherit;
}

.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-3 {
  display: inline-block;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6 [data-element='title'] {
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-6 [data-element='title'] > div > div > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6:hover {
  cursor: pointer;
}

.emotion-6:hover [data-element='title'] {
  -webkit-transform: translateX(5px);
  -ms-transform: translateX(5px);
  transform: translateX(5px);
}

.emotion-5 {
  margin-bottom: -2px;
  color: #879ba6;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  display: inline-block;
  color: rgba(135,155,166,0.7);
  text-transform: uppercase;
  font-weight: 500;
  font-size: 0.88rem;
  overflow: hidden;
  text-overflow: ellipsis;
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    font-size: 0.8rem;
  }
}

<div
  className="emotion-7"
>
  <div
    className="emotion-6"
  >
    <div
      data-element="title"
    >
      <div
        className="emotion-3"
      >
        <div
          className="emotion-2"
        >
          <div
            className="emotion-1"
            style={Object {}}
          >
            <div
              className="emotion-0"
            >
              Title
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="emotion-5"
    >
      <i
        className="emotion-4 Drycon Drycon-chevron-down"
      />
    </div>
  </div>
</div>
`;

exports[`Collapsible matches snapshot when it is open 1`] = `
.emotion-4 {
  color: inherit;
}

.emotion-1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-3 {
  display: inline-block;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6 [data-element='title'] {
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-6 [data-element='title'] > div > div > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-6:hover {
  cursor: pointer;
}

.emotion-6:hover [data-element='title'] {
  -webkit-transform: translateX(5px);
  -ms-transform: translateX(5px);
  transform: translateX(5px);
}

.emotion-5 {
  margin-bottom: -2px;
  color: #879ba6;
}

.emotion-7 {
  padding-top: 8px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  display: inline-block;
  color: rgba(135,155,166,0.7);
  text-transform: uppercase;
  font-weight: 500;
  font-size: 0.88rem;
  overflow: hidden;
  text-overflow: ellipsis;
}

@media only screen and (max-width:768px) {
  .emotion-0 {
    font-size: 0.8rem;
  }
}

<div
  className="emotion-8"
>
  <div
    className="emotion-6"
  >
    <div
      data-element="title"
    >
      <div
        className="emotion-3"
      >
        <div
          className="emotion-2"
        >
          <div
            className="emotion-1"
            style={Object {}}
          >
            <div
              className="emotion-0"
            >
              Title
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="emotion-5"
    >
      <i
        className="emotion-4 Drycon Drycon-chevron-up"
      />
    </div>
  </div>
  <div
    className="emotion-7"
  >
    Collapsible content
  </div>
</div>
`;

exports[`Collapsible matches snapshot when the title is a node 1`] = `
.emotion-5 {
  color: inherit;
}

.emotion-1 {
  color: #172b4e;
}

.emotion-2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-4 {
  display: inline-block;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-7 [data-element='title'] {
  -webkit-transition: all 0.15s ease-in-out;
  transition: all 0.15s ease-in-out;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-7 [data-element='title'] > div > div > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-7:hover {
  cursor: pointer;
}

.emotion-7:hover [data-element='title'] {
  -webkit-transform: translateX(5px);
  -ms-transform: translateX(5px);
  transform: translateX(5px);
}

.emotion-6 {
  margin-bottom: -2px;
  color: #879ba6;
}

.emotion-8 {
  padding-top: 8px;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-0 {
  display: inline;
  color: #172b4e;
  font-size: 1rem;
}

<div
  className="emotion-9"
>
  <div
    className="emotion-7"
  >
    <div
      data-element="title"
    >
      <div
        className="emotion-4"
      >
        <div
          className="emotion-3"
        >
          <div
            className="emotion-2"
            style={Object {}}
          >
            <div
              className="emotion-1"
            >
              <span
                className="emotion-0"
              >
                The title
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="emotion-6"
    >
      <i
        className="emotion-5 Drycon Drycon-chevron-up"
      />
    </div>
  </div>
  <div
    className="emotion-8"
  >
    Collapsible content
  </div>
</div>
`;
